### DISCUSSION TITLE
Trip-dependent loading time at the depot for CVRP_TW

### DISCUSSION THREAD

**Discussion by jrjrhan** [google/or-tools#2884] [Routing (and legacy CP) questions]
**What language and solver does this apply to?**
All, Python, Java, C#, C++ / CP-SAT, Routing, Linear Solver
Python

**Describe the problem you are trying to solve.**

Hi,

Thank you for the amazing tool for routing problem.

I'm currently implementing CVRP with Time Windows and Multi-Trip, and have been able to combine cvrp_tw and cvrp_reload for this combination for the open data (Solomon Data). Thanks, again!
But, in order to compare the result with the exisitng papers, I need to consider a loding time at the depot, which is defined as trip-dependent value rather than a constant. 
That is, given that each customer has their own service time, the loading time at the depot is set to 0.2 * the sum of the service times spent in the trip. So the loading time at the depot is depenent to which customers and how many customers are visited in that trip. 

Any help is appreciated.
Thanks!

**Comment by Mizux** [comment]
did you try to add a constraint on the slack var of the vehicle start node ?

1. Create a "ServiceTime" dimension i.e. for each node return the service time which imply
`serviceTime_dime.CumulVar(routing.End(vehicle_idx)) == sum of the service time along the route`

2. in you time dimension try to add something like that
```python
solver = routing.solver()
start = routing.Start(vehicle_idx)
end = routing.End(vehicle_idx)
solver.Add(time_dim.SlackVar(start) >= 0.2 * serviceTime_dim.CumulVar(end))
```

**Comment by jrjrhan** [comment]
Hi Mizux,

Thanks for the response.
I'm sort of new at using OR-tools, so please bear with me!

Based on your comments, I added the following function first:

```py
    def create_servicetime_evaluator(data):
        _servicetime = data['service_time']
    
        def servicetime_evaluator(manager, from_node):
            return _servicetime[manager.IndexToNode(from_node)]
    
        return servicetime_evaluator
```
And then, in function of add_time_window_constraints(routing, manager, data, time_evaluator, servicetime_evalutor),
I've added the following:
```py
    servicetime = 'ServiceTime'
    max_servicetime = data['vehicle_max_time']
    routing.AddDimension(
        servicetime_evaluator,
        max_servicetime,
        max_servicetime,
        True,
        servicetime)

    servicetime_dimension = routing.GetDimensionOrDie(servicetime)
    solver = routing.solver()
    for vehicle_id in range(data['num_vehicles']):
        start = routing.Start(vehicle_id)
        end = routing.End(vehicle_id)
        solver.Add(time_dimension.SlackVar(start) >= 0.2 * servicetime_dimension.CumulVar(end))
```
'TypeError: unsupported operand type(s) for *: 'float' and 'IntVar' happens at '0.2 * servicetime_dimension.CumulVar(end)'.

Plus, since I'm dealing with multi-trip based on cvrp_reload.py, I'm concerned whether the above 'for' loop is fine that way or the loop has to be made for each trip rather than each vehicle index. 

Thank you!

**Reply by Mizux** [reply]
First
```python    
routing.AddDimension(
        servicetime_evaluator,
        max_servicetime,
        max_servicetime,
        True,
        servicetime)
```
I guess here you need a `servicetime_evaluator_index` i.e.
```python
def create_servicetime_evaluator(data):
  _servicetime = data['service_time']
  def servicetime_evaluator(manager, from_node):
    return _servicetime[manager.IndexToNode(from_node)]
  return servicetime_evaluator

servicetime_evaluator_index = routing.RegisterUnaryTransitCallback(create_servicetime_evaluator(data))
routing.AddDimension(
        servicetime_evaluator_index,
```

Second, your issue, my bad, we don't support floating coefficient so let's try to use
`solver.Add(10 * time_dimension.SlackVar(start) >= 2 * servicetime_dimension.CumulVar(end))`
